### PR TITLE
Add pull secret into API

### PR DESCRIPTION
**Pull request by jim-minter**
This PR requires #251 to merge first and will need to be rebased.  Nonetheless review would be appreciated as otherwise I think this is pretty ready.

Idea is:
* in development, RP still uses PULL_SECRET env var
* in int and prod, RP will accept an optional pull secret from the end user.  It will augment that with an ACR token (that functionality is in #251).

### CODE REVIEW

File: pkg/backend/openshiftcluster/create.go [hunk 1]
```diff
@@ -124,6 +125,16 @@ func (m *Manager) Create(ctx context.Context) error {
 		},
 	}
 
+	pullSecret, err := pullsecret.Merge(os.Getenv("PULL_SECRET"), string(m.doc.OpenShiftCluster.Properties.ClusterProfile.PullSecret))
```
Comment (JackQuincy) [inline pkg/backend/openshiftcluster/create.go]: So this env pull secret will be there for dev getting the secrets for pulling in for repositories we need to setup the cluster.  What is the story in Prod/INT will it be on the customer to have a pull secret for every repo we use or will we have a core set we add onto their pull secret?

File: pkg/api/v20191231preview/openshiftcluster.go [hunk 2]
```diff
@@ -81,6 +81,9 @@ const (
 
 // ClusterProfile represents a cluster profile.
 type ClusterProfile struct {
+	// The pull secret for the cluster (immutable).
+	PullSecret string `json:"pullSecret,omitempty"`
```
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster.go]: @amanohar proposed API change.  To begin with, PullSecret is write-only and immutable.  Hopefully we can make it write-only + mutable later.

**Comment by jim-minter** [comment]
@mjudeikis or @asalkeld please can you review (but don't merge) this today?

File: python/az/aro/azext_aro/custom.py [hunk 3]
```diff
@@ -25,6 +25,7 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
                vnet=None,
                vnet_resource_group_name=None,  # pylint: disable=unused-argument
                location=None,
+               pull_secret=None,
```
Comment (nilsanderselde) [inline python/az/aro/azext_aro/custom.py]: If this defaulted to `''` here, then you wouldn't need to short-circuit on line 72.

File: python/client/azure/mgmt/redhatopenshift/v2019_12_31_preview/models/_models.py [hunk 4]
```diff
@@ -196,13 +198,15 @@ class ClusterProfile(Model):
...
     _attribute_map = {
+        'pull_secret': {'key': 'pullSecret', 'type': 'str'},
         'domain': {'key': 'domain', 'type': 'str'},
         'version': {'key': 'version', 'type': 'str'},
         'resource_group_id': {'key': 'resourceGroupId', 'type': 'str'},
     }
 
     def __init__(self, **kwargs):
         super(ClusterProfile, self).__init__(**kwargs)
+        self.pull_secret = kwargs.get('pull_secret', None)
```
Comment (nilsanderselde) [inline python/client/azure/mgmt/redhatopenshift/v2019_12_31_preview/models/_models.py]: When calling `get`, the value for `default` defaults to `None`, so you don't need to pass it explicitly here (also applies below).

File: pkg/install/samples.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		_, err = i.samplescli.SamplesV1().Configs().Update(c)
+		return err
+	})
+}
+
+// disableOperatorHubSources disables operator hub sources if there's no
+// appropriate pull secret
+func (i *Installer) disableOperatorHubSources(ctx context.Context) error {
+	if _, ok := i.env.(env.Dev); !ok &&
+		i.doc.OpenShiftCluster.Properties.ClusterProfile.PullSecret != "" {
```
[the remark this answers is not in the document]
Comment (nilsanderselde) [reply pkg/install/samples.go]: I'm currently adding tests to this package, so I could take care of that in a future PR.

File: python/client/azure/mgmt/redhatopenshift/v2019_12_31_preview/models/_models.py (hunk 4 above)
Comment (jim-minter) [reply python/client/azure/mgmt/redhatopenshift/v2019_12_31_preview/models/_models.py]: this file is autogenerated ;-)

File: python/az/aro/azext_aro/custom.py (hunk 3 above)
Comment (jim-minter) [reply python/az/aro/azext_aro/custom.py]: I don't know if defaulting to None is a requirement of the az code which wraps us.

**Pull request merged**